### PR TITLE
[distributed] allow rpc timeout to be set from python for different rpc backends

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29068 [distributed] allow rpc timeout to be set from python for different rpc backends**
* #28392 [distributed] add support for rpc timeouts in process group agent

RPC timeouts should be set from python for all backends, not just for process group agent

Differential Revision: [D18148449](https://our.internmc.facebook.com/intern/diff/D18148449/)

### CODE REVIEW

File: torch/distributed/rpc/backend_registry.py [hunk 1]
```diff
@@ -45,6 +45,7 @@ def process_group_init_backend_handler(
     self_rank,
     worker_name_to_id,
     num_send_recv_threads,
```
Comment (xush6528) [inline torch/distributed/rpc/backend_registry.py]: Can we remove `num_send_recv_threads ` as it's not common to all RPC agents at the moment.

File: torch/distributed/rpc/backend_registry.py [hunk 2]
```diff
@@ -76,7 +77,7 @@ def process_group_init_backend_handler(
                 )
             )
         # TODO: add try-except and destroy _agent in all processes if any fails.
-        return dist.ProcessGroupAgent(self_name, group, num_send_recv_threads, kwargs["rpc_timeout"])
+        return dist.ProcessGroupAgent(self_name, group, num_send_recv_threads, rpc_timeout)
```
Comment (xush6528) [inline torch/distributed/rpc/backend_registry.py]:
```cpp
return dist.ProcessGroupAgent(self_name, group, kwargs["num_send_recv_threads"], rpc_timeout)
```

Comment (rohan-varma) [reply torch/distributed/rpc/backend_registry.py]: This reads a bit odd as we are passing args from `kwargs` before regular arguments. If we want to do this, then we should change the argument order in process group agent?

File: torch/distributed/rpc/backend_registry.py (hunk 1 above)
Comment (rohan-varma) [reply torch/distributed/rpc/backend_registry.py]: This is a function specifically for initializing PG agent though, wouldn't it be better to have the argument explicitly?

Comment (xush6528) [reply torch/distributed/rpc/backend_registry.py]: Yeah, we can make it explicit. I think I am only hoping this argument not showing up in the arg list of `init_model_parallel`.

I think it's good now.

**Comment by pietern** [comment]
This needs a rebase now that #27529 is merged.

See https://github.com/pytorch/pytorch/commit/b4df413712098ff93da694472d95896e62faf07d#diff-d0c3a068b64bd33cff798988a76be111.

**Comment by xush6528** [comment]
Just noticed #29336 covers this PR. I am trying to unlock myself there.

File: torch/distributed/rpc/backend_registry.py (hunk 1 above)
Comment (mrshenli) [reply torch/distributed/rpc/backend_registry.py]: We can move `num_send_recv_threads` to `backend_config` once we have updated the `init_model_parallel` API. 

+1 it's good for now

Comment (xush6528) [reply torch/distributed/rpc/backend_registry.py]: @mrshenli 
No worry here. `num_send_recv_threads ` will be put into the `RpcAgentOptions` in my coming PR.

**Comment by xush6528** [comment]
@rohan-varma  Can we close this PR, given #29341 is ready?

**Pull request closed without merging**